rclcpp_lifecycle/include/rclcpp_lifecycle/lifecycle_publisher.hpp

```diff
@@ -0,0 +1,171 @@
+// Copyright 2016 Open Source Robotics Foundation, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef RCLCPP__LIFECYCLE_PUBLISHER_HPP_
+#define RCLCPP__LIFECYCLE_PUBLISHER_HPP_
+
+#include "rclcpp/publisher.hpp"
+
+namespace rclcpp
+{
+namespace publisher
+{
+
+namespace lifecycle_interface
+{
+/**
+ * @brief base class with only
+ * pure virtual functions. A managed
+ * node can then deactivate or activate
+ * the publishing.
+ * It is more a convenient interface class
+ * than a necessary base class.
+ */
+class PublisherInterface
+{
+public:
+  virtual void on_activate() = 0;
+  virtual void on_deactivate() = 0;
+};
+}  // namespace lifecycle_interface
+
+/**
+ * @brief child class of rclcpp Publisher class.
+ * Overrides all publisher functions to check for
+ * enabled/disabled state.
+ */
+template<typename MessageT, typename Alloc = std::allocator<void>>
+class LifecyclePublisher : public rclcpp::publisher::Publisher<MessageT, Alloc>,
+  public lifecycle_interface::PublisherInterface
+{
+public:
+  using MessageAllocTraits = allocator::AllocRebind<MessageT, Alloc>;
+  using MessageAlloc = typename MessageAllocTraits::allocator_type;
+  using MessageDeleter = allocator::Deleter<MessageAlloc, MessageT>;
+  using MessageUniquePtr = std::unique_ptr<MessageT, MessageDeleter>;
+
+  LifecyclePublisher(
+      std::shared_ptr<rcl_node_t> node_handle,
+      std::string topic,
+      const rcl_publisher_options_t & publisher_options,
+      std::shared_ptr<MessageAlloc> allocator):
+    ::rclcpp::publisher::Publisher<MessageT, Alloc>(node_handle, topic, publisher_options, allocator),
+      enabled_(false)
+  {  }
+
+  ~LifecyclePublisher() {};
+
+  /**
+   * @brief briefly checks whether this publisher
+   * was enabled or disabled and forwards the message
+   * to the actual rclcp Publisher base class
+   */
+  virtual void
+  publish(std::unique_ptr<MessageT, MessageDeleter> & msg)
+  {
+    if (!enabled_)
+    {
+      printf("I push every message to /dev/null\n");
```

Comment:
I agree with @dirk-thomas. The life cycle is intended to still allow things to happen in the `inactive` state, such as changing the behaviour of the node via parameters, that could not be done while the node is running its normal behaviour. A particular use case is a real-time node, where only the `active` state needs to actually be real-time safe. You would put your nodes into the `inactive` state, re-configure them, then re-activate them and return to normal behaviour.

So the user needs some control over which bits of communication get cut off and which do not. How that should be done, I have only vague ideas and no real proposals. :)